src/WebdriverClassicDriver.php

```diff
@@ -1,21 +1,1328 @@
+        }
+
+        $result = $result->getValue();
+        if ($result === null) {
+            return null;
+        }
+
+        return rawurldecode($result);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getContent(): string
+    {
+        return $this->getWebDriver()->getPageSource();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getScreenshot(): string
+    {
+        return $this->getWebDriver()->takeScreenshot();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getWindowNames(): array
+    {
+        $origWindow = $this->getWebDriver()->getWindowHandle();
+
+        try {
+            $result = [];
+            foreach ($this->getWebDriver()->getWindowHandles() as $tempWindow) {
+                $this->getWebDriver()->switchTo()->window($tempWindow);
+                $result[] = $this->getWindowName();
+            }
+            return $result;
+        } finally {
+            $this->getWebDriver()->switchTo()->window($origWindow);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getWindowName(): string
+    {
+        $name = (string)$this->evaluateScript('window.name');
+
+        if ($name === '') {
+            $name = self::W3C_WINDOW_HANDLE_PREFIX . $this->getWebDriver()->getWindowHandle();
+        }
+
+        return $name;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findElementXpaths(
+        #[Language('XPath')]
+        $xpath
+    ): array {
+        $nodes = $this->getWebDriver()->findElements(WebDriverBy::xpath($xpath));
+
+        $elements = [];
+        foreach ($nodes as $i => $node) {
+            $elements[] = sprintf('(%s)[%d]', $xpath, $i + 1);
+        }
+
+        return $elements;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTagName(
+        #[Language('XPath')]
+        $xpath
+    ): string {
+        return $this->findElement($xpath)->getTagName();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getText(
+        #[Language('XPath')]
+        $xpath
+    ): string {
+        return str_replace(["\r", "\n"], ' ', $this->findElement($xpath)->getText());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getHtml(
+        #[Language('XPath')]
+        $xpath
+    ): string {
+        return $this->executeJsOnXpath($xpath, 'return arguments[0].innerHTML;');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getOuterHtml(
+        #[Language('XPath')]
+        $xpath
+    ): string {
+        return $this->executeJsOnXpath($xpath, 'return arguments[0].outerHTML;');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getAttribute(
+        #[Language('XPath')]
+        $xpath,
+        $name
+    ): ?string {
+        $escapedName = $this->jsonEncode($name, 'get attribute', 'attribute name');
+        $script = "return arguments[0].getAttribute($escapedName)";
```

Comment:
can't we use `$element->getAttribute($name)`, at least for W3C-compliant sessions instead of this need for escaping ? JSONWire had a weird behavior where `getAttribute` wads not really getting the attribute, but this does not seem to happen in W3C webdriver.

Reply:
take a look at https://github.com/oleg-andreyev/MinkPhpWebDriver/blob/058c254b04f9586d16fccc82ae2bc0c5e88c402f/src/WebDriver.php#L540-L558

Reply:
@oleg-andreyev I thought about that as well and decided against it because:
1. that takes at least 3 requests, mine only takes one
2. it requires explaining why the extra logic is needed
3. this approach is backward compatible (in older selenium)